src/main/java/com/mongodb/starter/repositories/MongoDBPersonRepository.java

```diff
@@ -0,0 +1,137 @@
+    private MongoCollection<Person> personCollection;
+
+    @PostConstruct
+    void init() {
+        personCollection = client.getDatabase("test").getCollection("persons", Person.class);
+    }
+
+    @Override
+    public Person save(Person person) {
+        person.setId(new ObjectId());
+        personCollection.insertOne(person);
+        return person;
+    }
+
+    @Override
+    public List<Person> saveAll(List<Person> persons) {
+        try (ClientSession clientSession = client.startSession()) {
+            return clientSession.withTransaction(() -> {
+                persons.forEach(p -> p.setId(new ObjectId()));
+                personCollection.insertMany(clientSession, persons);
+                return persons;
+            }, txnOptions);
+        }
+    }
+
+    @Override
+    public List<Person> findAll() {
+        return personCollection.find().into(new ArrayList<>());
+    }
+
+    @Override
+    public List<Person> findAll(List<String> ids) {
+        return personCollection.find(in("_id", mapToObjectIds(ids))).into(new ArrayList<>());
+    }
+
+    @Override
+    public Person findOne(String id) {
+        return personCollection.find(eq("_id", new ObjectId(id))).first();
+    }
+
+    @Override
+    public long count() {
+        return personCollection.countDocuments();
+    }
+
+    @Override
+    public long delete(String id) {
+        return personCollection.deleteOne(eq("_id", new ObjectId(id))).getDeletedCount();
+    }
+
+    @Override
+    public long delete(List<String> ids) {
+        try (ClientSession clientSession = client.startSession()) {
+            return clientSession.withTransaction(
+                    () -> personCollection.deleteMany(clientSession, in("_id", mapToObjectIds(ids))).getDeletedCount(),
+                    txnOptions);
+        }
+    }
+
+    @Override
+    public long deleteAll() {
+        return personCollection.deleteMany(new BsonDocument()).getDeletedCount();
+    }
+
+    @Override
+    public Person update(Person person) {
+        FindOneAndReplaceOptions options = new FindOneAndReplaceOptions();
+        options.returnDocument(ReturnDocument.AFTER);
+        return personCollection.findOneAndReplace(eq("_id", person.getId()), person, options);
+    }
+
+    @Override
+    public long update(List<Person> persons) {
+        List<WriteModel<Person>> writes = persons.stream()
+                                                 .map(p -> new ReplaceOneModel<>(eq("_id", p.getId()), p))
+                                                 .collect(Collectors.toList());
+        try (ClientSession clientSession = client.startSession()) {
+            return clientSession.withTransaction(
+                    () -> personCollection.bulkWrite(clientSession, writes).getModifiedCount(), txnOptions);
+        }
+    }
+
+    @Override
+    public double getAverageAge() {
+        List<Bson> pipeline = asList(group(new BsonNull(), avg("averageAge", "$age")), project(excludeId()));
```

Comment:
When you group does each group end up with an `_id`?  I was surprised of the need for the projection here.

Reply:
Yes indeed :-).
![image](https://user-images.githubusercontent.com/7153911/66862414-231e5e80-ef91-11e9-8796-3f04d1fe31a4.png)

Reply:
Resolved